spec/commands/deploy.spec.ts

```diff
@@ -0,0 +1,393 @@
+import { EditGraphQLError, REMOTE_SERVER_CONTRACT_STATUS_SUBSCRIPTION } from "../../src/services/app/edit-graphql.js";
+import { type Context } from "../../src/services/command/context.js";
+import * as prompt from "../../src/services/output/prompt.js";
+import { nockTestApps } from "../__support__/app.js";
+import { makeMockEditGraphQLSubscriptions } from "../__support__/edit-graphql.js";
+import { loginTestUser } from "../__support__/user.js";
+
+describe("deploy", () => {
+  let ctx: Context<typeof args>;
+
+  beforeEach(() => {
+    loginTestUser();
+    nockTestApps();
+
+    ctx = makeContext(args);
+  });
+
+  afterEach(() => {
+    ctx.abort();
+    expect(nock.pendingMocks()).toEqual([]);
+  });
+
+  it("does not try to deploy if local files are not up to date with remote", async () => {
+    vi.spyOn(prompt, "select").mockResolvedValueOnce(Action.CANCEL);
+
+    await makeSyncScenario({ localFiles: { "file.txt": "test" } });
+
+    await expectProcessExit(() => deploy(ctx));
+  });
+
+  it("does not try to deploy if any problems were detected and displays the problems", async () => {
+    vi.spyOn(prompt, "select").mockResolvedValue("0");
+
+    await makeSyncScenario({ localFiles: { ".gadget/": "" } });
+
+    const mockEditGraphQL = makeMockEditGraphQLSubscriptions();
+
+    await deploy(ctx);
+
+    const publishStatus = mockEditGraphQL.expectSubscription(REMOTE_SERVER_CONTRACT_STATUS_SUBSCRIPTION);
+
+    publishStatus.emitResult({
+      data: {
+        publishStatus: {
+          remoteFilesVersion: "1",
+          progress: "NOT_STARTED",
+          issues: [
+            {
+              severity: "Error",
+              message: "Add google keys for production",
+              node: undefined,
+            },
+          ],
+        },
+      },
+    });
+
+    expectStdout().toMatchInlineSnapshot(`
+      "╭──────────────────────────────────────────────────────────────────────╮
+      │                                                                      │
+      │         ggt v0.3.3                                                   │
+      │                                                                      │
+      │         App         test                                             │
+      │         Editor      https://test.gadget.app/edit                     │
+      │         Playground  https://test.gadget.app/api/graphql/playground   │
+      │         Docs        https://docs.gadget.dev/api/test                 │
+      │                                                                      │
+      │         Endpoints                                                    │
+      │           • https://test.gadget.app                                  │
+      │           • https://test--development.gadget.app                     │
+      │                                                                      │
+      ╰──────────────────────────────────────────────────────────────────────╯
```

Comment:
I love the use of boxen to make the startup message stand out. However, I'm not sure this message makes sense for `ggt deploy`.

My intent behind this message was "a bunch of relevant links to use while developing your app", which is what I expect users are about to do when they run `ggt sync`.

I don't plan on showing this message when users run `ggt sync --once` though, because I expect users aren't about to start developing their app -- they're just pushing/pulling some changes and moving on.

---

This is just my personal opinion though, let me know what you think. If you both feel differently, then feel free to keep this message around! @jcao49 @infiton

Reply:
Good point! I added this to keep it consistent with the startup message sync shows and maybe to confirm to the user that the app they're deploying is the app they expect. If they deploy and theres no message I wonder if they'll feel confident that its the right app? Maybe we just want a simpler startup message, thoughts @infiton ?

Reply:
agree with @scott-rc don't think we need the big box for every deploy

Reply:
Updated it to just state the app slug when the command first runs- less chonky but still informs them which app they're deploying!